# Internship Report
## Introduction 
This is a report of my internship journey in the EANBiT Bioinformatics and Incubation and Mentorship programme at ICIPE which began on 6/10/2021 to 31/01/2022. The aim of this programme is to have well trained bioinformatcians to support genomics research. The internship involves a one month rigorous training involving both technical and soft skills learning, after which works collaboratively to reproduces a scientific paper to enhance the skills learnt and also to prepare for the mini-project. The mini-project provides hands on experience and enhances knowledge and skill retention.The mini-project also involves weekly code reviews to review progress and to also be guided on the project.

## Objectives
To learn how to analyse genomics data using different programming languages and different bioinformatics tools
To collaborate and network effectively
To gain soft skills such communication and presentation skills
To complete the mini-project as scheduled




## Training and skills learnt
 ### Purpose Roadmap 
In the begining of the internship i learnt how to write a [purpose roadmap](url) to outline what i hoped to achieve to by the end of the four months of internship. This has kept me motivated and  focused on my goals. In retrospect, i have achieved most of the things that i had purposed to do.



 ### Using Git and Github
Git is a version control system that helps keep track of history of the the things one is working on
GitHub is a web based service for version control and online collaboration.
I used Git and GitHub in creating and managing our projects, documenting codes, writing to do lists and working collaboratively in the projects. [Here](url) is a list of activities that i was involved in during the four months.



### Unix overview
I have learnt a wide range of commands to navigate through the command line, from using simple commands to writing complex scripts in bash;the command line language.
Bash is useful in automating repetitive tasks and combining smaller tasks into larger ones.
i used bash extensively when reproducing the scientific paper and in the mini-project and this has made using bash easy.

### Using the HPC
Genomics analysis involve processing huge amounts of data that are too much for a single computer to handle. High performance computing(HPC) helps overcome this challenge. It helps in processing huge amount of data at high speeds and provide more storage.
During the internship i have had access to the HPC at ICIPE where i have done most of the project's work.
i have learnt various commands to navigate through the nodes and to run different tasks.



### Introduction to workflows and package management
Workflows are tools that are used to automate bioinformatics analysis and  are useful in enhancing reproducibility.
Workflows simplify pipeline development,handling software installations and versions and the fact that the can be ran on different computing platforms enables portability and sharing, hence reproducibility of bioinformatics analyses.
i have learnt about workflows such as nextflow and snakemake,package management tools such conda and containers such as docker and singularity.
Using Nextflow in our mini-project has enhanced my understanding of this workflow language. i have learnt how to write and edit nextflow scripts and running different pipelines. 

 
### Foundational knowledge in bioinformatics
Bioinformatics involves use of computational approaches to biological data to organize analyze and interprete such data. There is a vast range of bioinformatics tools, resources and databases that are available
Databases are structured collection of records stored in a computer system. Genomic databases store DNA or protein sequences as well as annotated information about these sequences.
Bioinformatics tools are programs used to analyze these sequences.
i have learnt about bioinformatics resources such NCBI and EMBL-EBI and bioinformatics databases and tools 

### Introduction to Genomics
i have learnt about the various sequencing technologies such as Sanger sequencing,Illumina sequencing, oxford nanopore technology, and also various NGS file formats



### Python programming
I have learnt to use jupyter notebooks which are useful in documenting the python codes.
The python skills that i have acquired include:
- Basic data types and operations
- Data structures list tuples and dictionaries
- Functions and classes 
- Reading files and importing modules
- Control statements



### Phylogenetics
I learnt various methods for constructing phylogenetic trees such as UPGMA, neighbour joining, maximum parsimony and maximum likehood methods and how to interprete phylogenetic trees. This knowledge was important in reproducing the scientific paper.

### Reproducing a scientific paper
We reproduced a scientfic paper: Global Outbreaks and spread of a chikungunya virus variant carrying mutations which may increase fitness for Aedes aegypti: Revelations from the 2016 Mandera Kenya Outbreak. 
[Here](url) is link to the repository for this work
i learnt how to use different phylogenetic tools such as tempest, mafft, phyml and raxml

### Journal club presentation
During the internship we have had biweekly journal club presentations where one presents a scientific paper on their area of interests.
This has had me develop a culture of reading scientific papers, and also enhanced my communication and presentation skills.

### Mini-project
Our mini-project was focused on assessing different variant calling pipelines identify great ones and expand where there are gaps.
We assessed the different pipelines and were to to automate fully the mbbu variant calling pipeline.
This project helped to understand Nextflow a workflow language, to use bash the commandline language extensively, to use the HPC, to use various bioinformatics tools, to collaborate effectively, it also helped me improve my communication and presentation skills.
[Here](url) is link to the repository for this mini-project


### Soft skills 
In addition to the extensive technical skills that i have learnt, this internship program also entailed learning essential soft skills which included:
    - Leadership skills
    - Communication and presentation  skills
    - Giving feedback and conflict resolution 
    - Career planning 
    - Writing competitive applications
 
 ## Conclusion
 This internship program, has equipped me with  the necessary skills and has laid a great foundation to further my career in bioinformatics.I am now more motivated and confident to pursue my career further.
 I accord profound gratitude to the entire bioinformatics team for their great dedication to teach, guide and mentor us through out the four months internship journey.
 
 ## Recommendations
 - To have the mini-project introduced earlier in the internship so that one familiarises with the topic and with the skills required before the actual time for starting the mini-project

